src/dynamic_programming/knapsack.md

```diff
@@ -0,0 +1,157 @@
+
+## 0-1 Knapsack
+
+### Explaination
+
+The known conditions in the example are; the weight of $i^{th}$ item $w_{i}$, value of $i^{th}$ item $v_{i}$, and the total capacity of the knapsack {W}.
+
+Set the dp state $f_{i, j}$ be the maximum total value the knapsack can carry with capacity $j$, when only the first $i$ items are considered.
+
+Consider the state transfer. Assuming that all states of first $i-1$ items have been processed, for the $i^{th}$ item;
+- When it is not put into the knapsack, the remaining capacity remains unchanged and total value does not change. Therefore, the maximum value in this case is $f_{i-1, j}$
+- When it is put into the knapsack, the remaining capacity decreases by $w_{i}$ and the total value increases by $v_{i}$,
+so the maximum value in this case is $f_{i-1, j-w_i} + v_i$
+
+From this we can derive the state transfer equation:
+
+$$f_{i, j} = \max(f_{i-1, j}, f_{i-1, j-w_i} + v_i)$$
+
+Further, as $f_{i}$ is only dependent on $f_{i-1}$, we can remove the first dimension. We obtain:
+
+$$f_{j} = \max(f_{j}, f_{j-w_i} + v_i)$$
+
+**It is important to remember and understand this transfer equation, because most of the transfer equations for knapsack problems are derived on this basis.**
+
+### Implementation
+
+Another thing to note is that it is easy to write **erroneous code** like this
+
+```.c++
+for (int i = 1; i <= n; i++)
+  for (int j = 0; j <= W-w[i]; j++)
+    f[j + w[i]] = max(f[j + w[i]], f[j] + v[i]);
+```
+
+What is wrong with this code? The enumeration order is wrong.
+
+Observing the code carefully, we can find that: for the currently processed item $i$ and the current state $f_{i,j}$, 
+when $j\geqslant w_{i}$, $f_{i,j}$ will be affected by $f_{i,j-w_{i}}$. 
+This is equivalent to being able to be put item $i$ into the backpack multiple times, which is not consistent with the question.
+(In fact, this is exactly the solution to the complete knapsack problem)
+
+To avoid this, we can change the order of enumeration from $W$ to $w_{i}$, so that the above error won't occour, because $f_{i, j}$ is always updated before $f_{i, j-w_i}$
+
+Therefore, the actual code is
+
+```.c++
+for (int i = 1; i <= n; i++)
+  for (int j = W; j >= w[i]; j--)
+    f[j] = max(f[j], f[j - w[i]] + v[i]);
+```
+
+## Complete Knapsack
+
+The complete knapsack model is similar to the 0-1 knapsack, the only difference from the 0-1 knapsack is that an item can be selected an unlimited number of times instead of only once.
+
+We can refer to the idea of 0-1 knapsack to define the state: $f_{i, j}$, the maximum value the knapsack can obtain using the first $i$ items with maximum capacity $j$.
+
+It should be noted that although the state definition is similar to that of a 0-1 knapsack, its state transfer equation is different from that of a 0-1 knapsack.
+
+### Explaination
+
+The trivial approach is, for the first $i$ items, enumerate how many each item is to be taken. The time complexity of this is $O(n^3)$.
```

Comment:
It should be  $O(n^2 W)$, I suppose? We should also add complexity in the previous and this section, which should be $O(n W)$.